lib/snapshot_logger/snapshot_logger.ex

```diff
@@ -0,0 +1,129 @@
+defmodule SnapshotLogger.SnapshotLogger do
+  @moduledoc """
+  Logs algorithm snapshots on a cadence.
+  """
+  use GenServer
+  require Logger
+  alias Alerts.Alert
+  alias AlertsViewer.DelayAlertAlgorithm
+  alias Routes.{Route, RouteStatsPubSub}
+
+  @cadence 5 * 60 * 1000
+
+  @spec start_link(any) :: :ignore | {:error, any} | {:ok, pid}
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl GenServer
+  def init(opts) do
+    delay_alert_algorithm_modules =
+      Application.get_env(:alerts_viewer, :delay_alert_algorithm_components)
+
+    bus_alerts = Alerts.subscribe() |> filtered_by_bus() |> filtered_by_delay_type()
+    bus_routes = Routes.all_bus_routes(Keyword.get(opts, :route_opts, []))
+    stats_by_route = RouteStatsPubSub.subscribe()
+    routes_with_current_alerts = Enum.filter(bus_routes, &delay_alert?(&1, bus_alerts))
+    schedule_logs()
+
+    {:ok,
+     %{
+       bus_routes: bus_routes,
+       stats_by_route: stats_by_route,
+       routes_with_current_alerts: routes_with_current_alerts,
+       delay_alert_algorithm_modules: delay_alert_algorithm_modules,
+       routes_with_recommended_alerts: []
+     }}
+  end
+
+  @impl GenServer
+  def handle_info({:stats_by_route, stats_by_route}, state) do
+    {:noreply, Map.put(state, :stats_by_route, stats_by_route)}
+  end
+
+  @impl GenServer
+  def handle_info({:alerts, alerts}, state) do
+    bus_alerts = filtered_by_bus(alerts) |> filtered_by_delay_type()
+
+    routes_with_current_alerts = Enum.filter(state.bus_routes, &delay_alert?(&1, bus_alerts))
+
+    state =
+      Map.put(
+        state,
+        :routes_with_current_alerts,
+        routes_with_current_alerts
+      )
+
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(:log, state) do
+    schedule_logs()
+
+    snapshot =
+      state.delay_alert_algorithm_modules
+      |> Enum.reduce(%{}, fn module, acc ->
+        snapshot_data = make_snapshot(module, state)
+        Map.put(acc, module_name(module), snapshot_data)
+      end)
+      |> Jason.encode_to_iodata!()
```

Comment:
thought: Cool, I wasn't familiar with the `encode_to_iodata!` function.

Reply:
I just happened to see it in the Jason documentation, it says it's a bit more efficient for our purposes.